uint32 handle
uint8 mode
string mode_name
bool valid